### DISCUSSION TITLE
Union of hashables is not hashable

### DISCUSSION THREAD

**Discussion by JanEricNitschke** [microsoft/pyright#8268] [Q&A] [answered]
I was just trying to simplify some hints in pandas-stubs where there are 4 Hashable bound typevars in a signature that are each only used once. https://github.com/pandas-dev/pandas-stubs/blob/main/pandas-stubs/core/reshape/concat.pyi#L28

I thought that that should be easily replaceable with just using Hashable directly but this gives me an error:

```
def func(a: Mapping[Hashable, pd.DataFrame]) -> None:
    ...

value = {1: pd.DataFrame(), "a": pd.DataFrame()}
func(value)
```

```
error: Argument of type "dict[int | str, DataFrame]" cannot be assigned to parameter "a" of type "Mapping[Hashable, DataFrame]" in function "func"
    "dict[int | str, DataFrame]" is incompatible with "Mapping[Hashable, DataFrame]"
      Type parameter "_KT@Mapping" is invariant, but "int | str" is not the same as "Hashable" (reportArgumentType)
```

Is there something i can do to resolve this or do i need to keep the four TypeVars that are each only used once here.

**Comment by erictraut** [comment]
Pyright's behavior here is correct. This has nothing to do with unions. The first type parameter in the `Mapping` class is invariant, so `Mapping[int, DataFrame]` is not assignable to `Mapping[Hashable, DataFrame]`. These are the basic rules of class subtyping and variance.

Here's a simplified example that demonstrates the issue:

```python
from typing import Hashable, Mapping

def func(value: Mapping[int, str]):
    x: Mapping[Hashable, str] = value # Type Error
```

Not surprisingly, mypy also generates an error with this code.

**Reply by JanEricNitschke** [reply]
Ok, I understand. Then what would be the correct way to handle this? Explicitly declaring the input dict as taking a hashable key before calling the function? Although that would put more burden on the caller. Or just keeping with these otherwise unnecessary type vars?

**Reply by erictraut** [reply, marked as answer]
You could use a type variable with an upper bound of `Hashable`.

```python
import pandas as pd
from typing import Hashable, Mapping

def func[T: Hashable](a: Mapping[T, pd.DataFrame]) -> None: ...

value = {1: pd.DataFrame(), "a": pd.DataFrame()}
func(value)
```

**Reply by JanEricNitschke** [reply]
Ok yeah, that is what is currently being used in the pandas stubs. The signature has 4 independence hashable bound typevars. It just felt wasteful with each of them being only used once.
